include/dfu_client.h

```diff
@@ -0,0 +1,191 @@
+ *          This feedback is useful when, for example, a fault fragment was received or the application
+ *          was unable to process a fimrware fragment.
+ *
+ * @param[in] dfu 		The DFU instance.
+ * @param[in] event     The event.
+ * @param[in] status    Event status, can be 0 or an errno value.
+ *
+ * @retval 0 indicates the event was handled successfully.
+ *         Else, the application failed to handle the event.
+ */
+typedef int (*dfu_client_event_handler_t)(struct dfu_client_object *dfu, enum dfu_client_evt event,
+					u32_t status);
+
+/**@brief Firmware download object describing the state of download*/
+struct dfu_client_object {
+	char resp_buf[CONFIG_NRF_DFU_HTTP_MAX_RESPONSE_SIZE];
+	char req_buf[CONFIG_NRF_DFU_HTTP_MAX_REQUEST_SIZE];
+	char *fragment;
+	int fragment_size;
+
+	int fd;			/**< Transport file descriptor.
+					  *  If negative the transport is disconnected.
+					  */
+	int firmware_size;	/**< Total size of firmware being downloaded.
+						 * If negative, download is in progress.
+						 * If zero, the size is unknown.
+						 */
+	volatile int download_size;	/**< Size of firmware being downloaded. */
+	volatile int status;		/**< Status of transfer. Will be one of \ref dfu_client_status. */
+	const char *host;			   /**< The server that hosts the firmwares. */
+	const char *resource;          /**< Resource to be downloaded. */
+
+	const dfu_client_event_handler_t callback; /**< Event handler.
+												* Shall not be NULL.
+												*/
+};
+
+/**@brief Initialize firmware upgrade object for a given host and resource.
+ *
+ * @details The server to connect to for firmware download is identified by
+ * the host. The caller of this method must register an event handler in
+ * callback field of the instance.
+ *
+ * @param[inout] dfu The DFU instance. Shall not be NULL.
+ *		     The target, host, resource and callback fields should be
+ *		     correctly initialized in the object instance.
+ *		     The fd, status, fragment, fragment_size, download_size and
+ *           firmware_size fields are out parameters.
+ *
+ * @retval 0 if the procedure succeeded, else, -1.
+ * @note If this method fails, no other APIs shall be called for the instance.
+ */
+int dfu_client_init(struct dfu_client_object *dfu);
+
+/**@brief Establish connection to the server.
+ *
+ * @details The server to connect to for firmware download is identified by
+ * the host field of the dfu object. The host field is expected to be hostname
+ * that can be resolved using DNS to an IP address.
+ *
+ * @param[in] dfu The DFU instance.
+ *
+ * @retval 0 if the procedure request succeeded, else, -1.
+ *
+ * @note This is a blocking call.
+ * @note Do not initiate a @ref dfu_client_download if this procedure fails.
+ */
+int dfu_client_connect(struct dfu_client_object *dfu);
+
+/**@brief Start download of firmware.
+ *
+ * @breif This is a blocking call used to trigger download of a firmware image identified
```

Comment:
```@details```
But by the way, the usage of ```@details``` here is redundant. See how it is done in other libraries.
The paragraphs is treated as detailed description automatically. So leaving just empty line after ```@brief``` is enough.